oecophylla/taxonomy/metaphlan2.py

```diff
@@ -0,0 +1,287 @@
+
+    Raises
+    ------
+    IOError
+        If the file cannot be read.
+    """
+    clades = {}
+    try:
+        file = open(filename, 'r')
+        for line in file:
+            if line.startswith('gi|'):
+                type_ids = 'gi'
+                accession = (line.split('\t')[0]).split('|')[1]
+            elif line.startswith('GeneID:'):
+                type_ids = 'GeneID'
+                accession = (line.split('\t')[0]).split(':')[1]
+            elif line.startswith('NC_'):
+                type_ids = 'NC'
+                accession = line.split('\t')[0]
+            else:
+                type_ids = None
+                accession = None
+
+            if (type_ids is not None) and (accession is not None):
+                clade = line.split("clade': '")[1].split("'")[0]
+                if clade not in clades:
+                    clades[clade] = {}
+                if type_ids not in clades[clade]:
+                    clades[clade][type_ids] = {}
+                clades[clade][type_ids][accession] = True
+
+        for clade in clades:
+            for type_id in clades[clade]:
+                clades[clade][type_id] = set(clades[clade][type_id].keys())
+
+        file.close()
+        return clades
+
+    except IOError:
+        raise IOError('Cannot read file "%s"' % filename)
+
+
+def _read_ncbitaxonomy_file(filename):
+    """ A generic function to read an NCBI taxonomy file, which is delimited
+    by '\t|\t?'.
+
+    Parameters
+    ----------
+    filename : str
+        Path to a file from an NCBI taxonomy dump.
+
+    Returns
+    -------
+    A dict. Key = first field of file, Value = second field of file.
+
+    Raises
+    ------
+    IOError
+        If the file cannot be read.
+    ValueError
+        If IDs of entries cannot be converted into int.
+    """
+    entries = {}
+    try:
+        file = open(filename, 'r')
+        for line in file:
+            fields = list(map(str.strip, line.split('|')))
+            try:
+                entries[int(fields[0])] = int(fields[1])
+            except ValueError:
+                file.close()
+                raise ValueError("cannot convert entry IDs (%s, %s) to int."
+                                 % (fields[0], fields[1]))
+
+        file.close()
+        return entries
+
+    except IOError:
+        raise IOError('Cannot read file "%s"' % filename)
+
+
+def read_ncbi_merged(filename):
+    """ Reads NCBI's merged.dmp file and returns a dict of old and merged IDs.
+
+    Parameters
+    ----------
+    filename : str
+        Path to the filename 'merged.dmp' of NCBI's taxonomy.
+
+    Returns
+    -------
+    A dict, where keys are old IDs and their values are the new merged IDs.
+
+    Raises
+    ------
+    IOError
+        If the file cannot be read.
+    ValueError
+        If IDs of old or merged nodes cannot be converted into int.
+    """
+    return _read_ncbitaxonomy_file(filename)
+
+
+def update_taxids(input, updatedTaxids):
```

Comment:
I am not sure how robust `merges.dmp` is. Is it necessary to check whether TaxIDs in `updatedTaxids` do exist in the current TaxID map?

Reply:
I don't fully understand.
merges.dmp has two columns (old and new taxid). If a taxid in my collection exists in the dump, than it is time to update it to the newer value, otherwise I leave the taxid unchanged. I do relay on the correctness of the merges.dmp file itself and it MUST be the matching one to the taxonomy dump that you want to use.

Reply:
Makes sense!